operator/v1/types_vsphere_problem_detector.go

```diff
@@ -0,0 +1,66 @@
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// VSphereProblemDetector is used to manage and report health of Openshift
+// cluster running in vSphere environment.
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// VSphereProblemDetector objects allows configuration and health reporting of
+// Vsphere problem detector operator.
+type VSphereProblemDetector struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec holds user settable values for configuration
+	// +kubebuilder:validation:Required
+	// +required
+	Spec VSphereProblemDetectorSpec `json:"spec"`
+
+	// status holds observed values from the cluster. They may not be overridden.
+	// +optional
+	Status VSphereProblemDetectorStatus `json:"status"`
+}
+
+// VSphereProblemDetectorSpec is the desired configuration of vSphere problem detector operator.
+type VSphereProblemDetectorSpec struct {
+	OperatorSpec `json:",inline"`
+}
+
+// VSphereHealthCheckStatus is the status of vSphere problem detector operator and
+// health of Openshift and vSphere integration.
+type VSphereProblemDetectorStatus struct {
+	// List of checks and their status
+	Checks         []VSphereHealthCheck `json:"vSphereHealthCheck,omitempty"`
+	OperatorStatus `json:",inline"`
+}
+
+type VSphereHealthCheck struct {
+	Name          string             `json:"name"`
+	Result        VSphereCheckResult `json:"vSphereCheckResult"`
+	Message       string             `json:"message"`
+	LastCheckTime metav1.Time        `json:"lastCheckTime,omitempty"`
+}
```

Comment:
this looks a lot like a condition, doesn't it?

Reply:
it does yeah. The idea is report bunch of healthcheck status from vsphere API. The difference is - operatorcondition reports condition of the operator itself whereas `VSphereHealthCheck` will report status of checks.

Reply:
I think LastCheckTime should be global, in `Status` - all checks will have the same timestamp. Or not?

Reply:
after discussing offline, moved this to global value in status field.

Reply:
Can we add a new `CheckTime`-like field in `VSphereHealthCheck`  and get rid of `LastCheckTime` (in the .Status)? Then  `VSphereProblemDetectorStatus.Checks[len(VSphereProblemDetectorStatus.Checks)-1]` will contain the "last check time".

IMO every `VSphereHealthCheck` should contain the time it was performed.

Reply:
> Then VSphereProblemDetectorStatus.Checks[len(VSphereProblemDetectorStatus.Checks)-1] will contain the "last check time".

No, the `status.checks` is not history of checks, it's list of individual checks - like ["CheckNodeUUID", "CheckFilePermissions", "CheckTasksPermissions", ...]. Only metrics will have history of all checks.

And we don't want to have different exp. backoff for each check - it could be quite confusing to users ("CheckNodeUUID" succeeds, so it is tested every 8 hours, but "CheckTasksPermissions" failed, so it will be re-tested in couple of minutes). Therefore the last check time will be the same for all checks.

Reply:
Yes, so this was discussed in slack and when I looked at the code, it made sense to just keep one value.  We can revisit this in future if needed.